src/proto/Ethereum.proto

```diff
@@ -3,6 +3,58 @@ syntax = "proto3";
 package TW.Ethereum.Proto;
 option java_package = "wallet.core.jni.proto";
 
+// Transaction (transfer, smart contract call, ...)
+message Transaction {
+    // Native coin transfer transaction
+    message Transfer {
+        // Amount to send in wei (256-bit number)
+        bytes amount = 1;
+
+        // Optional payload data
+        bytes data = 2;
+    }
+
+    // ERC20 token transfer transaction
+    message ERC20Transfer {
+        string to = 1;
+
+        // Amount to send (256-bit number)
+        bytes amount = 2;
+    }
+
+    // ERC20 approve transaction
+    message ERC20Approve {
+        string spender = 1;
+
+        // Amount to send (256-bit number)
+        bytes amount = 2;
+    }
+
+    // ERC721 NFT transfer transaction
+    message ERC721Transfer {
+        string from = 1;
+
+        string to = 2;
+
+        // ID of the token (256-bit number)
+        bytes token_id = 3;
+    }
+
+    // Generic smart contract transaction
+    message ContractGeneric {
+        // Contract call payload data
+        bytes data = 1;
+    }
+
+    oneof transaction_oneof {
+        Transfer transaction_transfer = 1;
```

Comment:
```suggestion
        Transfer transfer = 1;
```

no need to add prefix `transaction_`? , looks verbose

Reply:
I've copied the "oneof ENTITY_oneof { type1 ENTITY_type1 = 1; type2 ENTITY_type2 = 2; ... }" pattern seen in some other .proto file.

Reply:
Removed